File: backend/app/api/v1/endpoints/waitlists.py
from typing import Annotated

from fastapi import APIRouter, Depends, status
from sqlalchemy.orm import Session

from app.api.deps import get_current_student_id
from app.db.session import get_db
from app.modules.registration.publishers import RedisAvailabilityPublisher
from app.modules.waitlists.schemas import WaitlistCreate, WaitlistDeleteResponse, WaitlistItem
from app.modules.waitlists.service import WaitlistService

router = APIRouter()
DbSession = Annotated[Session, Depends(get_db)]


@router.get("/me", response_model=list[WaitlistItem])
def list_my_waitlists(
    student_id: Annotated[int, Depends(get_current_student_id)],
    db: DbSession,
) -> list[WaitlistItem]:
    return WaitlistService(db).list_current(student_id)


@router.post("", response_model=WaitlistItem, status_code=status.HTTP_201_CREATED)
def join_waitlist(
    payload: WaitlistCreate,
    student_id: Annotated[int, Depends(get_current_student_id)],
    db: DbSession,
) -> WaitlistItem:
    return WaitlistService(db, RedisAvailabilityPublisher(db)).join(student_id, payload)


@router.delete("/{waitlist_entry_id}", response_model=WaitlistDeleteResponse)
def cancel_waitlist(
    waitlist_entry_id: int,
    student_id: Annotated[int, Depends(get_current_student_id)],
    db: DbSession,
) -> WaitlistDeleteResponse:
    return WaitlistService(db, RedisAvailabilityPublisher(db)).cancel(student_id, waitlist_entry_id)

File: frontend/src/api/waitlists.ts
import client from './client'

export interface WaitlistItem {
  waitlist_entry_id: number
  section_id: number
  position: number
  status: string
}

export async function getMyWaitlists(): Promise<WaitlistItem[]> {
  const res = await client.get('/api/v1/waitlists/me')
  return res.data
}

export async function joinWaitlist(section_id: number): Promise<WaitlistItem> {
  const res = await client.post('/api/v1/waitlists', { section_id })
  return res.data
}

export async function cancelWaitlist(
  waitlistEntryId: number,
): Promise<{ status: string; waitlist_entry_id: number }> {
  const res = await client.delete(`/api/v1/waitlists/${waitlistEntryId}`)
  return res.data
}
